### ISSUE TITLE
`.cuda`,`.device` integratiom

### ISSUE THREAD

**Issue by jjjkkkjjj** [jjjkkkjjj/pytorch_SSD#1]

**Comment by jjjkkkjjj** [comment]
```python
        start = time.time()
        if self.ignore:
            bboxes, linds, flags = self.ignore(bboxes, linds, flags)
        print(time.time()-start)
        start = time.time()
        if self.augmentation:
            img, bboxes, linds, flags = self.augmentation(img, bboxes, linds, flags)
        print(time.time() - start)
        start = time.time()
        if self.transform:
            img, bboxes, linds, flags = self.transform(img, bboxes, linds, flags)
        print(time.time() - start)
        start = time.time()
        if self.target_transform:
            bboxes, linds, flags = self.target_transform(bboxes, linds, flags)
        print(time.time() - start)
```

```bash
1.9073486328125e-05
0.0057544708251953125
0.011091947555541992
6.747245788574219e-05
1.621246337890625e-05
0.0545656681060791
0.0033698081970214844
5.9604644775390625e-05
1.52587890625e-05
0.003098726272583008
0.0006515979766845703
4.029273986816406e-05
1.4781951904296875e-05
0.007569789886474609
0.0007729530334472656
3.981590270996094e-05
1.52587890625e-05
0.005670309066772461
0.0006487369537353516
4.744529724121094e-05
1.1920928955078125e-05
0.03446769714355469
0.0010843276977539062
4.076957702636719e-05
1.33514404296875e-05
0.005401134490966797
0.0006537437438964844
3.600120544433594e-05
1.239776611328125e-05
0.014462471008300781
0.0007925033569335938
4.220008850097656e-05
1.2159347534179688e-05
0.003934383392333984
0.0004489421844482422
3.552436828613281e-05
1.621246337890625e-05
0.002662181854248047
0.0006313323974609375
3.3855438232421875e-05
1.5497207641601562e-05
0.01498270034790039
0.00086212158203125
5.3882598876953125e-05
1.239776611328125e-05
0.003958702087402344
0.0004603862762451172
3.743171691894531e-05
1.1444091796875e-05
0.05571627616882324
0.0011408329010009766
4.124641418457031e-05
1.1682510375976562e-05
0.0022933483123779297
0.0010199546813964844
3.743171691894531e-05
1.2636184692382812e-05
0.05581068992614746
0.0011107921600341797
4.00543212890625e-05
1.239776611328125e-05
0.018535852432250977
0.0011286735534667969
3.981590270996094e-05
1.33514404296875e-05
0.002362489700317383
0.0004456043243408203
3.62396240234375e-05
1.1682510375976562e-05
0.01344442367553711
0.0009765625
4.458427429199219e-05
1.3828277587890625e-05
0.002716541290283203
0.0005748271942138672
3.933906555175781e-05
1.2874603271484375e-05
0.014225959777832031
0.0008821487426757812
4.267692565917969e-05
1.4543533325195312e-05
0.002471446990966797
0.0004730224609375
3.695487976074219e-05
1.4543533325195312e-05
0.03576397895812988
0.001184701919555664
4.744529724121094e-05
1.3828277587890625e-05
0.0026509761810302734
0.00041866302490234375
3.2901763916015625e-05
1.0013580322265625e-05
0.015186309814453125
0.0008594989776611328
3.838539123535156e-05
1.1682510375976562e-05
0.007502555847167969
0.0005044937133789062
4.172325134277344e-05

...

```

**Comment by jjjkkkjjj** [comment]
Compared to [ssd.pytorch](https://github.com/amdegroot/ssd.pytorch), iteration time was almost same except for iter0.

pytorch_SSD

>Training... Epoch: 1, Iter: 1,	 [32/21503	 (0%)]	Loss: 25.586593, Loc Loss: 3.392233, Conf Loss: 22.194361	Iter time: 0.6660
Training... Epoch: 1, Iter: 10,	 [320/21503	 (1%)]	Loss: 15.622597, Loc Loss: 2.794373, Conf Loss: 12.828224	Iter time: 0.1698
Training... Epoch: 1, Iter: 20,	 [640/21503	 (3%)]	Loss: 16.242902, Loc Loss: 3.165571, Conf Loss: 13.077331	Iter time: 0.1709
Training... Epoch: 1, Iter: 29,	 [928/21503	 (4%)]	Loss: 14.366256, Loc Loss: 2.820415, Conf Loss: 11.545841	Iter time: 0.1793

ssd.pytorch

>iter 0 || Loss: 26.5115, Loc: 3.3145, Conf: 23.1969 || timer: 0.1829 sec.
iter 10 || Loss: 15.6610, Loc: 2.9267, Conf: 12.7343 || timer: 0.1781 sec.
iter 20 || Loss: 17.0699, Loc: 2.8916, Conf: 14.1783 || timer: 0.1729 sec.
iter 30 || Loss: 16.3119, Loc: 2.9908, Conf: 13.3211 || timer: 0.1783 sec.

**Comment by jjjkkkjjj** [comment]
however, `update_iteration` was too slow!!!

```python
                start_ = time.time()
                # update train
                self.log_manager.update_iteration(self.model, epoch, _iteration + 1, batch_num=len(images), data_num=len(train_loader.dataset),
                                                  iter_per_epoch=len(train_loader), loclossval=locloss.item(), conflossval=confloss.item(), iter_time=end-start)
                print()
                print(time.time() - start_)
```

**Comment by jjjkkkjjj** [comment]
`enumeate` may have caused to this problem??

```python
        start_ = time.time()
        for epoch in range(1, epochs + 1):
            if self.log_manager.isFinish:
                break

            for _iteration, (images, targets) in enumerate(train_loader):
                self.optimizer.zero_grad()

                if self.gpu:
                    images = images.cuda()
                    targets = [target.cuda() for target in targets]
                start = time.time()

                print()
                print(time.time() - start_)

                # set variable
                # images.requires_grad = True
                # targets.requires_grad = True
                pos_indicator, predicts, gts = self.model.learn(images, targets)

                confloss, locloss = self.loss_func(pos_indicator, predicts, gts)
                loss = confloss + self.loss_func.alpha * locloss
                loss.backward()  # calculate gradient for value with requires_grad=True, shortly back propagation
                #print(self.model.feature_layers.convRL1_1.conv.weight.grad)
                #print(self.model.localization_layers.conv_loc_1.weight.grad)
                self.optimizer.step()
                if self.scheduler:
                    self.scheduler.step()

                end = time.time()
                # update train
                self.log_manager.update_iteration(self.model, epoch, _iteration + 1, batch_num=len(images), data_num=len(train_loader.dataset),
                                                  iter_per_epoch=len(train_loader), loclossval=locloss.item(), conflossval=confloss.item(), iter_time=end-start)
                start_ = time.time()
                """
                too slow...
                if evaluator and self.log_manager.now_iteration % evaluator.iteration_interval:
                    self.model.eval()
                    print(evaluator(self.model))
                    self.model.train()
                """
                if self.log_manager.isFinish:
                    break
```

>Training... Epoch: 1, Iter: 1,	 [32/21503	 (0%)]	Loss: 26.361246, Loc Loss: 3.281082, Conf Loss: 23.080164	Iter time: 0.6553
0.6686105728149414
Training... Epoch: 1, Iter: 2,	 [64/21503	 (0%)]	Loss: 24.843592, Loc Loss: 3.317637, Conf Loss: 21.525955	Iter time: 0.4240
0.5093734264373779
Training... Epoch: 1, Iter: 3,	 [96/21503	 (0%)]	Loss: 24.072326, Loc Loss: 3.107326, Conf Loss: 20.965000	Iter time: 0.1719
0.5080349445343018

**Comment by jjjkkkjjj** [comment]
The reason was `num_workers` in `DataLoader`... :(
